analyze.module

```diff
@@ -25,4 +29,232 @@ function analyze_theme() {
+        if (method_exists($entity, 'getEntityType') && $type = $entity->getEntityType()) {
+          if (method_exists($type, 'getBundleOf') && $entity_type = $type->getBundleOf()) {
+            $is_bundle_form = TRUE;
+            $bundle = $entity->id();
+          }
+        }
+      }
+    }
+
+    if ($is_bundle_form) {
+      analyze_add_settings_form($form, $entity_type, $bundle);
+      $form_state->set('entity_type', $entity_type);
+      $form_state->set('bundle', $bundle);
+    }
+  }
+}
+
+/**
+ * Helper to add Analyze settings to a form.
+ *
+ * @param mixed[] $form
+ *   The form.
+ * @param string $entity_type
+ *   The entity type the settings are for.
+ * @param string $bundle
+ *   The bundle the settings are for.
+ */
+function analyze_add_settings_form(array &$form, string $entity_type, string $bundle): void {
+  if (\Drupal::currentUser()->hasPermission('administer analyze')) {
+    $config = \Drupal::config('analyze.settings')->get('status');
+
+    $form['analyze'] = [
+      '#type' => 'details',
+      '#title' => t('Analyze settings'),
+      '#description' => t('Enable Analyze reports for this entity.'),
+      '#parents' => ['analyze'],
+      '#open' => !empty($config[$entity_type][$bundle]),
+    ];
+
+    foreach (\Drupal::service('plugin.manager.analyze')->getDefinitions() as $id => $definition) {
+      $form['analyze'][$id] = [
+        '#type' => 'checkbox',
+        '#title' => $definition['label'],
+        '#default_value' => isset($config[$entity_type][$bundle][$id]),
+        '#parents' => ['analyze', $id],
+      ];
+    }
+
+    if (isset($form['actions']['submit']['#submit'])) {
+      $form['actions']['submit']['#submit'][] = 'analyze_add_settings_form_submit';
+    }
+    elseif (isset($form['#submit'])) {
+      $form['#submit'][] = 'analyze_add_settings_form_submit';
+    }
+  }
+}
+
+/**
+ * Store the Analyze settings, if they have been set.
+ *
+ * @param mixed[] $form
+ *   The Drupal form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ */
+function analyze_add_settings_form_submit(array &$form, FormStateInterface $form_state): void {
+  if ($values = $form_state->cleanValues()->getValue('analyze')) {
+    $config = \Drupal::configFactory()->getEditable('analyze.settings')->get('status');
+    $entity_type = $form_state->get('entity_type');
+    $bundle = $form_state->get('bundle');
+
+    // Clear any old settings for this bundle.
+    $config[$entity_type][$bundle] = [];
+
+    foreach ($values as $plugin_id => $status) {
+      if ($status) {
+        $config[$entity_type][$bundle][$plugin_id] = TRUE;
+      }
+    }
+
+    \Drupal::configFactory()->getEditable('analyze.settings')
+      ->set('status', $config)
+      ->save();
+  }
+}
+
+/**
+ * Helper to get all entities with a canonical route.
+ *
+ * @return string[]
+ *   An array of entity id => canonical route name.
+ */
+function analyze_get_entity_types(): array {
```

Comment:
If you move to derivers, I would suggest moving this into a service that can be injected. Or if it's only used b the derivers, a trait.

Reply:
It was there because I could use the trait in the module: now it's a service I've removed it and updated the code.